File: arena-backend/src/modules/file/s3.service.ts
import { Inject, Injectable, OnModuleInit } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { 
  S3Client, 
  PutObjectCommand, 
  GetObjectCommand, 
  DeleteObjectCommand,
  HeadObjectCommand,
  HeadBucketCommand,
  CreateBucketCommand,
  PutBucketAclCommand
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import Redis from "ioredis";
import { REDIS_CLIENT } from "src/redis/redis.constants";

@Injectable()
export class S3Service implements OnModuleInit {
  private static readonly PRESIGNED_URL_EXPIRES_IN = 86400; // 24시간
  private static readonly CACHE_TTL = 82800; // 23시간
  private static readonly CACHE_REFRESH_THRESHOLD = 3600; // 1시간
  private static readonly S3_BASE_URL = process.env.S3_BASE_URL || "http://localhost:14566";

  static getFileUrl(bucket: string, storageKey: string): string {
    return `${S3Service.S3_BASE_URL}/${bucket}/${storageKey}`;
  }

  private readonly s3Client: S3Client;
  private readonly publicBucket: string;
  private readonly privateBucket: string;

  constructor(
    private readonly configService: ConfigService,
    @Inject(REDIS_CLIENT)
    private readonly redis: Redis,
  ) {
    this.s3Client = new S3Client({
      region: this.configService.get<string>("AWS_REGION") || "us-east-1",
      endpoint: this.configService.get<string>("AWS_ENDPOINT"),
      credentials: {
        accessKeyId: this.configService.get<string>("AWS_ACCESS_KEY_ID") || "test",
        secretAccessKey: this.configService.get<string>("AWS_SECRET_ACCESS_KEY") || "test",
      },
      forcePathStyle: true,
    });

    this.publicBucket = this.configService.get<string>("S3_PUBLIC_BUCKET") || "arena-files-public";
    this.privateBucket = this.configService.get<string>("S3_PRIVATE_BUCKET") || "arena-files-private";
  }

  async onModuleInit() {
    await this.ensureBucketExists(this.publicBucket, true);
    await this.ensureBucketExists(this.privateBucket, false);
  }

  private async ensureBucketExists(bucketName: string, isPublic: boolean): Promise<void> {
    try {
      await this.s3Client.send(new HeadBucketCommand({ Bucket: bucketName }));
      console.log(`✅ Bucket exists: ${bucketName}`);
    } catch (error: any) {
      if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
        try {
          await this.s3Client.send(new CreateBucketCommand({ Bucket: bucketName }));
          console.log(`✅ Created bucket: ${bucketName}`);
          
          if (isPublic) {
            await this.s3Client.send(new PutBucketAclCommand({
              Bucket: bucketName,
              ACL: 'public-read',
            }));
            console.log(`✅ Set bucket public: ${bucketName}`);
          }
        } catch (createError) {
          console.error(`❌ Failed to create bucket: ${bucketName}`, createError);
          throw createError;
        }
      } else {
        console.error(`❌ Failed to check bucket: ${bucketName}`, error);
        throw error;
      }
    }
  }

  async getPresignedUploadUrlPublic(
    key: string, 
    mimeType: string, 
    expiresIn: number = 3600
  ): Promise<string> {
    const command = new PutObjectCommand({
      Bucket: this.publicBucket,
      Key: key,
      ContentType: mimeType,
      ACL: 'public-read',
    });

    return await getSignedUrl(this.s3Client, command, { expiresIn });
  }

  async getPresignedUploadUrlPrivate(
    key: string, 
    mimeType: string, 
    expiresIn: number = 3600
  ): Promise<string> {
    const command = new PutObjectCommand({
      Bucket: this.privateBucket,
      Key: key,
      ContentType: mimeType,
    });

    return await getSignedUrl(this.s3Client, command, { expiresIn });
  }

  async getPresignedDownloadUrl(
    bucket: 'public' | 'private',
    key: string,
  ): Promise<string> {
    // Public은 캐싱 불필요 (static URL)
    if (bucket === 'public') {
      return S3Service.getFileUrl(this.publicBucket, key);
    }

    // Private는 캐싱
    const cacheKey = `presigned:${bucket}:${key}`;
    
    // 1. 캐시 확인
    const cached = await this.redis.get(cacheKey);
    if (cached) {
      const ttl = await this.redis.ttl(cacheKey);
      
      // TTL이 충분히 남았으면 재사용
      if (ttl > S3Service.CACHE_REFRESH_THRESHOLD) {
        return cached;
      }
    }

    // 2. 새로 생성
    const url = await this.generatePresignedDownloadUrl(bucket, key);

    // 3. 캐싱
    await this.redis.set(cacheKey, url, "EX", S3Service.CACHE_TTL);

    return url;
  }

  private async generatePresignedDownloadUrl(
    bucket: 'public' | 'private',
    key: string,
  ): Promise<string> {
    const bucketName = bucket === 'public' ? this.publicBucket : this.privateBucket;
    
    const command = new GetObjectCommand({
      Bucket: bucketName,
      Key: key,
    });

    return await getSignedUrl(
      this.s3Client, 
      command, 
      { expiresIn: S3Service.PRESIGNED_URL_EXPIRES_IN }
    );
  }

  async delete(bucket: 'public' | 'private', key: string): Promise<void> {
    const bucketName = bucket === 'public' ? this.publicBucket : this.privateBucket;
    
    const command = new DeleteObjectCommand({
      Bucket: bucketName,
      Key: key,
    });

    await this.s3Client.send(command);

    // 캐시 삭제
    const cacheKey = `presigned:${bucket}:${key}`;
    await this.redis.del(cacheKey);
  }

  async exists(bucket: 'public' | 'private', key: string): Promise<boolean> {
    const bucketName = bucket === 'public' ? this.publicBucket : this.privateBucket;
    
    try {
      const command = new HeadObjectCommand({
        Bucket: bucketName,
        Key: key,
      });
      await this.s3Client.send(command);
      return true;
    } catch (error: any) {
      if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
        return false;
      }
      throw error;
    }
  }

  async getMetadata(
    bucket: 'public' | 'private',
    key: string
  ): Promise<{ contentType: string; contentLength: number } | null> {
    const bucketName = bucket === 'public' ? this.publicBucket : this.privateBucket;
    
    try {
      const command = new HeadObjectCommand({
        Bucket: bucketName,
        Key: key,
      });
      const response = await this.s3Client.send(command);
      
      return {
        contentType: response.ContentType || 'application/octet-stream',
        contentLength: response.ContentLength || 0,
      };
    } catch (error: any) {
      if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
        return null;
      }
      throw error;
    }
  }
}

File: legacy/backend/src/main/java/app/sandori/arena/api/domain/file/S3Service.java
package app.sandori.arena.api.domain.file;

import jakarta.annotation.PostConstruct;
import java.time.Duration;
import java.util.concurrent.TimeUnit;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.BucketAlreadyExistsException;
import software.amazon.awssdk.services.s3.model.BucketAlreadyOwnedByYouException;
import software.amazon.awssdk.services.s3.model.CreateBucketRequest;
import software.amazon.awssdk.services.s3.model.DeleteObjectRequest;
import software.amazon.awssdk.services.s3.model.HeadBucketRequest;
import software.amazon.awssdk.services.s3.model.HeadObjectRequest;
import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
import software.amazon.awssdk.services.s3.model.NoSuchBucketException;
import software.amazon.awssdk.services.s3.model.NoSuchKeyException;
import software.amazon.awssdk.services.s3.model.ObjectCannedACL;
import software.amazon.awssdk.services.s3.model.PutBucketAclRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;
import software.amazon.awssdk.services.s3.presigner.model.PresignedPutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest;

@Slf4j
@Service
public class S3Service {

    private static final int PRESIGNED_URL_EXPIRES_IN = 86400;
    private static final long CACHE_TTL = 82800;
    private static final long CACHE_REFRESH_THRESHOLD = 3600;

    private static String s3BaseUrl;

    private final S3Client s3Client;
    private final S3Presigner s3Presigner;
    private final StringRedisTemplate redisTemplate;
    private final String publicBucket;
    private final String privateBucket;

    public S3Service(S3Client s3Client,
                     S3Presigner s3Presigner,
                     StringRedisTemplate redisTemplate,
                     @Value("${aws.s3.public-bucket}") String publicBucket,
                     @Value("${aws.s3.private-bucket}") String privateBucket,
                     @Value("${aws.s3.base-url}") String baseUrl) {
        this.s3Client = s3Client;
        this.s3Presigner = s3Presigner;
        this.redisTemplate = redisTemplate;
        this.publicBucket = publicBucket;
        this.privateBucket = privateBucket;
        s3BaseUrl = baseUrl;
    }

    public static String getFileUrl(String bucket, String storageKey) {
        return s3BaseUrl + "/" + bucket + "/" + storageKey;
    }

    @PostConstruct
    void init() {
        ensureBucketExists(publicBucket, true);
        ensureBucketExists(privateBucket, false);
    }

    private void ensureBucketExists(String bucketName, boolean isPublic) {
        try {
            s3Client.headBucket(HeadBucketRequest.builder().bucket(bucketName).build());
            log.info("Bucket exists: {}", bucketName);
        } catch (NoSuchBucketException e) {
            try {
                s3Client.createBucket(CreateBucketRequest.builder().bucket(bucketName).build());
                log.info("Created bucket: {}", bucketName);

                if (isPublic) {
                    s3Client.putBucketAcl(PutBucketAclRequest.builder()
                            .bucket(bucketName)
                            .acl("public-read")
                            .build());
                    log.info("Set bucket public: {}", bucketName);
                }
            } catch (BucketAlreadyExistsException | BucketAlreadyOwnedByYouException ex) {
                log.info("Bucket already exists: {}", bucketName);
            } catch (Exception ex) {
                log.error("Failed to create bucket: {}", bucketName, ex);
            }
        } catch (Exception e) {
            log.error("Failed to check bucket: {}", bucketName, e);
        }
    }

    public String getPresignedUploadUrlPublic(String key, String mimeType, int expiresIn) {
        PutObjectRequest putRequest = PutObjectRequest.builder()
                .bucket(publicBucket)
                .key(key)
                .contentType(mimeType)
                .acl(ObjectCannedACL.PUBLIC_READ)
                .build();

        PutObjectPresignRequest presignRequest = PutObjectPresignRequest.builder()
                .signatureDuration(Duration.ofSeconds(expiresIn))
                .putObjectRequest(putRequest)
                .build();

        PresignedPutObjectRequest presigned = s3Presigner.presignPutObject(presignRequest);
        return presigned.url().toString();
    }

    public String getPresignedUploadUrlPrivate(String key, String mimeType, int expiresIn) {
        PutObjectRequest putRequest = PutObjectRequest.builder()
                .bucket(privateBucket)
                .key(key)
                .contentType(mimeType)
                .build();

        PutObjectPresignRequest presignRequest = PutObjectPresignRequest.builder()
                .signatureDuration(Duration.ofSeconds(expiresIn))
                .putObjectRequest(putRequest)
                .build();

        PresignedPutObjectRequest presigned = s3Presigner.presignPutObject(presignRequest);
        return presigned.url().toString();
    }

    public void delete(String bucket, String key) {
        String bucketName = "public".equals(bucket) ? publicBucket : privateBucket;
        s3Client.deleteObject(DeleteObjectRequest.builder()
                .bucket(bucketName)
                .key(key)
                .build());

        String cacheKey = "presigned:" + bucket + ":" + key;
        redisTemplate.delete(cacheKey);
    }

    public record S3Metadata(String contentType, long contentLength) {
    }

    public S3Metadata getMetadata(String bucket, String key) {
        String bucketName = "public".equals(bucket) ? publicBucket : privateBucket;
        try {
            HeadObjectResponse response = s3Client.headObject(HeadObjectRequest.builder()
                    .bucket(bucketName)
                    .key(key)
                    .build());
            return new S3Metadata(
                    response.contentType() != null ? response.contentType() : "application/octet-stream",
                    response.contentLength() != null ? response.contentLength() : 0
            );
        } catch (NoSuchKeyException e) {
            return null;
        }
    }

    private String resolveBucketName(String bucket) {
        return "public".equals(bucket) ? publicBucket : privateBucket;
    }
}
